### PR TITLE
Made the application work again

### PR DESCRIPTION
**Pull request by Ry-DS**
Migrated to rowak's new Nanoleaf API, which seemed to have fixed everything.

Things that are broken:
- Token for the shapes isn't saved, always prompted to save again
- Spotify login should've been saved
- Audio is quite delayed from the effect

~~The latter is likely fixable by reworking how the spotify api is used.~~
You also can't add devices via IP (I hard coded mine in, gotta remove later)

### CODE REVIEW

**Comment by jaxcksn** [comment]
So the issue was the nanoleaf API or the spotify API (I was suspecting the Nanoleaf API)

**Comment by Ry-DS** [comment]
> So the issue was the nanoleaf API or the spotify API (I was suspecting the Nanoleaf API)

After looking into it more, the main issue is using effects to try update panels in real time, as they aren't really designed for that use case. 

This should be used instead: https://github.com/rowak/nanoleaf-api#external-streaming

It uses a pure stream of UDP packets which will be a lot faster than using json effects (what's being used currently)

**Comment by jaxcksn** [comment]
Ah, see that was my plan for the new program. I disliked using Java  for the UDP packets.

This is a good foundation for the new code base (in which I'm basically gonna skip using a third party nanaleaf api, try to reduce overhead)

**Comment by Ry-DS** [comment]
> Ah, see that was my plan for the new program. I disliked using Java for the UDP packets.
> 
> This is a good foundation for the new code base (in which I'm basically gonna skip using a third party nanaleaf api, try to reduce overhead)

Are you planning on not using Java/third party apis purely for speed/functionality? Or for a learning experience? If for the former I highly recommend just sticking with what you have now (and using the UDP implementation linked) since switching to C++/raw processing won't give a significant enough speedup to justify losing this nice UI you built. The task isn't computationally expensive (more of a network problem) which Java can handle just fine if done correctly.

**Comment by jaxcksn** [comment]
The UI is actually one of the reasons I want to switch. Working with JavaFX is a pain, also I think that importing a whole third party library for the API is overkill if I'm only using 4 endpoints for the entire thing.

I haven't decided on C++, I might do C# instead but we'll have to see.

**Comment by Ry-DS** [comment]
> The UI is actually one of the reasons I want to switch. Working with JavaFX is a pain, also I think that importing a whole third party library for the API is overkill if I'm only using 4 endpoints for the entire thing.
> 
> I haven't decided on C++, I might do C# instead but we'll have to see.

Yup that sounds fair. Hope this PR is of some help as it makes the current app work with the modern Nanoleaf devices (shapes, elements etc)

**Comment by jaxcksn** [comment]
This PR is very helpful. I'm not gonna merge it quite yet, as I'm starting the new code base in a few days but I appreciate the help.

**Comment by jaxcksn** [comment]
Going to have Java be in a separate branch from the master, which will contain the new code base.

File: src/main/java/dev/jaxcksn/nanoleafMusic/controllers/ConnectToDevice.java [hunk 1]
```diff
@@ -90,13 +93,14 @@ public String toString() {
...
-    private List<AuroraMetadata> findNanoleaf() throws IOException {
-        List<AuroraMetadata> auroras = new ArrayList<>();
+    private List<NanoleafDeviceMeta> findNanoleaf() throws IOException {
+        List<NanoleafDeviceMeta> auroras = new ArrayList<>();
 
         try {
             Service service = Service.fromName("_nanoleafapi._tcp");
             Query query = Query.createFor(service, Domain.LOCAL);
             Set<Instance> instances = query.runOnce();
+            auroras.add(new NanoleafDeviceMeta("192.168.0.20", 16021 ,"xxx","Shapes.Hexagon"));
```
Comment (Ry-DS) [inline src/main/java/dev/jaxcksn/nanoleafMusic/controllers/ConnectToDevice.java:103]: Forgot to remove this before merging. Might need to fix this quickly

Comment (jaxcksn) [reply src/main/java/dev/jaxcksn/nanoleafMusic/controllers/ConnectToDevice.java:103]: Open an issue and then make a PR for it real quick to the Java branch

**Pull request merged**